### DISCUSSION TITLE
How to attach mutliple services to a Builder?

### DISCUSSION THREAD

**Discussion by weezy20** [hyperium/hyper#2691] [Q&A] [answered]
`let server = Server::bind(&addr).serve(service1).serve(service2);` doesn't works as I was expecting since `serve` consumes the builder. I would like to attach multiple services. I create services from functions using the `make_service_fn` method and as I understand it, all it does is take a `Request<Body>` and return a `Response`. Am I misunderstanding something here? The lack of documentation on these matters doesn't make it easier to figure it out.

**Comment by seanmonstar** [comment, marked as answer]
A server only uses 1 service. When a connection is accepted, it calls the "make service", so you can do any connection-specific setup, and then return the `Service` which will handle all requests on that connection.

Do you have different services that are based on routes? Then you'll need to do routing in a single service. You could use something like [routerify](https://crates.io/crates/routerify) to do so. Or look at the [Echo guide](https://hyper.rs/guides/server/echo/) that shows routing without any other library.
